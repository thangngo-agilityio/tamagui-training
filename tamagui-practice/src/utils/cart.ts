// Types

import { ICartItem } from '@/types';

export const calculateTotalPrice = (cartItems: ICartItem[] = []) =>
  cartItems.reduce((acc, cartItem) => {
    const { product, quantity = 1 } = cartItem || {};
    const { price = 0 } = product || {};

    const itemPrice = price * quantity;

    return acc + itemPrice;
  }, 0);

export const calculateTotalQuantity = (cartItems: ICartItem[] = []) =>
  cartItems.reduce((acc, cartItem) => {
    const { quantity = 0 } = cartItem || {};

    return acc + quantity;
  }, 0);
