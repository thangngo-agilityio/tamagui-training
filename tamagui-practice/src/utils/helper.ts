export const formatAmountNumber = (value: string): string => {
  if (!value) {
    return value;
  }

  if (Number.isNaN(parseFloat(value))) {
    return '';
  }

  const dotIndex = value.indexOf('.');
  const newValue =
    dotIndex > -1
      ? value.slice(0, dotIndex).replaceAll(',', '')
      : value.replaceAll(',', '');

  const newValueFormat = newValue.replace(/\B(?=(\d{3})+(?!\d))/g, ',');

  return newValueFormat;
};
