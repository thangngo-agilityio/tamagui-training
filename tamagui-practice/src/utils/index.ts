export * from './images';
export * from './pagination';
export * from './paramUrl';
export * from './helper';
export * from './cart';
export * from './format';
export * from './validators';
