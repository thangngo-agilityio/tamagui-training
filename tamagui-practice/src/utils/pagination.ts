// Constants
import { DOTS } from '@/constants';

// Types
import { FormatPaginationParams, PaginationTableType } from '@/types';

export const formatNumberButton = (numberOfPage: number): string[] =>
  Array.from({ length: numberOfPage }).map((_, index: number): string =>
    (index + 1).toString(),
  );

export const formatPageArray = ({
  totalPage = 0,
  currentPage = 0,
}: PaginationTableType): string[] => {
  // If the number of pages is less than or equal to 5, display all page buttons
  if (totalPage <= 5) {
    return formatNumberButton(totalPage);
  }

  // If the current page is near the start (pages 1 to 3),
  // Display the first 3 pages and then DOTS and the last page
  if (currentPage <= 3) {
    return [...formatNumberButton(3), DOTS, totalPage.toString()];
  }

  // If the current page is near the end (within the last 3 pages),
  // display the first page, DOTS, and the last 3 pages
  if (currentPage >= totalPage - 2) {
    return ['1', DOTS, ...formatNumberButton(totalPage).splice(-3)];
  }

  // If the current page is in the middle,
  // Display the first page, DOTS, currentPage, DOTS, and the last page
  return ['1', DOTS, currentPage.toString(), DOTS, totalPage.toString()];
};

export const formatPagination = ({
  totalCount,
  limit,
  currentPage,
}: FormatPaginationParams): string[] => {
  const totalPage = Math.ceil(totalCount / limit);

  return formatPageArray({ totalPage, currentPage });
};
