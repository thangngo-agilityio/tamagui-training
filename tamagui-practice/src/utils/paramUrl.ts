export const getSearchParams = (
  searchParams: URLSearchParams,
): Record<string, string> => {
  const queryParam: Record<string, string> = {};

  searchParams.forEach((value, key) => {
    queryParam[key] = value;
  });

  return queryParam;
};

export const updateSearchParams = (
  searchParams: URLSearchParams,
  key: string,
  value: string,
): URLSearchParams => {
  const params = new URLSearchParams(searchParams);

  if (value) {
    params.set(key, value);
  } else {
    params.delete(key);
  }

  return params;
};

export const formatUrlWithQuery = (
  endpoint: string,
  query: Record<string, string | string[] | number | number[] | undefined>,
): string => {
  const params = new URLSearchParams(
    Object.entries(query)
      .filter(([, value]) => value !== undefined)
      .flatMap(([key, value]) =>
        Array.isArray(value)
          ? value.map((val) => [key, String(val)])
          : [[key, String(value)]],
      ),
  );

  return params.toString() ? `${endpoint}?${params.toString()}` : endpoint;
};
