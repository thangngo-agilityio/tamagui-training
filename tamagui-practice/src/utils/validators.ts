export const isEnableSubmitButton = (
  requiredFields: string[],
  dirtyFields: string[],
  errors: Record<string, unknown>,
): boolean => {
  const isMatchAllRequiredFields: boolean = requiredFields.every((field) =>
    dirtyFields.includes(field),
  );

  return isMatchAllRequiredFields && errors && !Object.keys(errors).length;
};
