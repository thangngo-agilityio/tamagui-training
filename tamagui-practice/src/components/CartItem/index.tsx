import Image from 'next/image';
import { Box, Flex, Heading, Text, VStack } from '@chakra-ui/react';
import { useCallback } from 'react';

// Utils
import { formatAmountNumber } from '@/utils';

// Icons
import { DeleteIcon } from '@/icons';

type TCartItem = {
  id: string;
  title: string;
  image: string[];
  description: string;
  price: number;
  quantity: number;
  onRemoveItem?: (id: string) => void;
};

const CartItem = ({
  id,
  title,
  image,
  description,
  price,
  quantity,
  onRemoveItem,
}: TCartItem) => {
  const handleRemoveItem = useCallback(() => {
    onRemoveItem && onRemoveItem(id);
  }, [id, onRemoveItem]);
  return (
    <Flex flexDir={{ base: 'column', lg: 'row' }} gap="30px">
      <Flex flex={1}>
        <Image
          width={347}
          height={290}
          src={image[0]}
          alt={title}
          style={{
            objectFit: 'cover',
            width: '100%',
            height: '100%',
            borderRadius: '8px',
          }}
        />
      </Flex>
      <Flex flex={4} flexDir="column">
        <Heading
          size={{ base: 'size4xl', lg: 'size6xl' }}
          variant="quinary"
          mb="5px"
        >
          {title}
        </Heading>
        <Text
          size={{ base: 'textMd', lg: 'text2Xl' }}
          variant="senary"
          mb="30px"
        >
          {description}
        </Text>

        <VStack mb="30px" alignItems="flex-start">
          <Heading variant="productTitle" size="size2xl">
            Size:
          </Heading>
          <Flex flexDir="row" gap="5px">
            <Text variant="senary" size="text2Xl">
              Length-34cm,
            </Text>
            <Text variant="tertiary" size="text2Xl">
              Width-56cm
            </Text>
          </Flex>
        </VStack>

        <Box w="50%" borderTopWidth="1px" borderColor="border.300" mb="10px">
          <Heading size="size9xl" variant="quinary" py="10px">
            N{formatAmountNumber(price?.toString())}
          </Heading>
        </Box>

        <Flex gap="10px">
          <Heading variant="productTitle" size="size2xl">
            Quantity:
          </Heading>
          <Text variant="tertiary" size="text2Xl">
            {quantity}
          </Text>
        </Flex>
      </Flex>

      <Flex justifyContent="center" alignItems="center">
        <Box height="fit-content" onClick={handleRemoveItem} cursor="pointer">
          <DeleteIcon />
        </Box>
      </Flex>
    </Flex>
  );
};

export default CartItem;
