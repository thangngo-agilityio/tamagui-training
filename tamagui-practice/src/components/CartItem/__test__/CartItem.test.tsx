import React from 'react';
import { formatAmountNumber } from '@/utils';
import { fireEvent, render, screen } from '@testing-library/react';
import CartItem from '../index';
import '@testing-library/jest-dom';

// Mock the formatAmountNumber utility function
jest.mock('@/utils', () => ({
  formatAmountNumber: jest.fn((amount) => amount),
}));

// Mock the DeleteIcon component
jest.mock('@/icons', () => ({
  DeleteIcon: () => <svg data-testid="delete-icon" />,
}));

describe('CartItem() CartItem method', () => {
  const defaultProps = {
    id: '1',
    title: 'Sample Item',
    image: ['https://sample-image.jpg'],
    description: 'This is a sample item description.',
    price: 1000,
    quantity: 2,
    onRemoveItem: jest.fn(),
  };

  it('should match snapshot for CartItem', () => {
    const { container } = render(<CartItem {...defaultProps} />);

    expect(container).toMatchSnapshot();
  });

  describe('Happy Path', () => {
    it('should render the CartItem with all provided props', () => {
      // Render the CartItem component
      render(<CartItem {...defaultProps} />);

      // Check if the title is rendered
      expect(screen.getByText('Sample Item')).toBeInTheDocument();

      // Check if the description is rendered
      expect(
        screen.getByText('This is a sample item description.'),
      ).toBeInTheDocument();

      // Check if the image is rendered
      const image = screen.getByAltText('Sample Item');
      expect(image).toBeInTheDocument();
      expect(image).toHaveAttribute(
        'src',
        '/_next/image?url=https%3A%2F%2Fsample-image.jpg&w=750&q=75',
      );

      // Check if the price is formatted and rendered
      expect(formatAmountNumber).toHaveBeenCalledWith('1000');
      expect(screen.getByText('N1000')).toBeInTheDocument();

      // Check if the quantity is rendered
      expect(screen.getByText('2')).toBeInTheDocument();
    });

    it('should call onRemoveItem when the delete icon is clicked', () => {
      // Render the CartItem component
      render(<CartItem {...defaultProps} />);

      // Click the delete icon
      fireEvent.click(screen.getByTestId('delete-icon'));

      // Check if onRemoveItem is called with the correct id
      expect(defaultProps.onRemoveItem).toHaveBeenCalledWith('1');
    });
  });

  describe('Edge Cases', () => {
    it('should handle missing onRemoveItem gracefully', () => {
      // Render the CartItem component without onRemoveItem
      render(<CartItem {...defaultProps} onRemoveItem={undefined} />);

      // Click the delete icon
      fireEvent.click(screen.getByTestId('delete-icon'));

      // Check if onRemoveItem is not called
      expect(defaultProps.onRemoveItem).not.toHaveBeenCalled();
    });

    it('should render with a price of 0', () => {
      // Render the CartItem component with a price of 0
      render(<CartItem {...defaultProps} price={0} />);

      // Check if the price is formatted and rendered
      expect(formatAmountNumber).toHaveBeenCalledWith('0');
      expect(screen.getByText('N0')).toBeInTheDocument();
    });
  });
});
