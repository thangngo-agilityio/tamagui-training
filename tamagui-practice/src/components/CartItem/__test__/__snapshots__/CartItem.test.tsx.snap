// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem() CartItem method should match snapshot for CartItem 1`] = `
<div>
  <div
    class="css-fnno8d"
  >
    <div
      class="css-13pmxen"
    >
      <img
        alt="Sample Item"
        data-nimg="1"
        decoding="async"
        height="290"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2Fsample-image.jpg&w=750&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fsample-image.jpg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Fsample-image.jpg&w=750&q=75 2x"
        style="color: transparent; object-fit: cover; width: 100%; height: 100%; border-radius: 8px;"
        width="347"
      />
    </div>
    <div
      class="css-1lvotsk"
    >
      <h2
        class="chakra-heading css-r7q7qr"
      >
        Sample Item
      </h2>
      <p
        class="chakra-text css-1xn3edc"
      >
        This is a sample item description.
      </p>
      <div
        class="chakra-stack css-1ih7iax"
      >
        <h2
          class="chakra-heading css-0"
        >
          Size:
        </h2>
        <div
          class="css-tx8ton"
        >
          <p
            class="chakra-text css-0"
          >
            Length-34cm,
          </p>
          <p
            class="chakra-text css-0"
          >
            Width-56cm
          </p>
        </div>
      </div>
      <div
        class="css-1u85xi0"
      >
        <h2
          class="chakra-heading css-9452eu"
        >
          N
          1000
        </h2>
      </div>
      <div
        class="css-i5q2k0"
      >
        <h2
          class="chakra-heading css-0"
        >
          Quantity:
        </h2>
        <p
          class="chakra-text css-0"
        >
          2
        </p>
      </div>
    </div>
    <div
      class="css-1n2mv2k"
    >
      <div
        class="css-ua50fe"
      >
        <svg
          data-testid="delete-icon"
        />
      </div>
    </div>
  </div>
</div>
`;
